Sources/AsyncHTTPClient/AsyncAwait/TransactionBody.swift

```diff
@@ -0,0 +1,65 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021-2022 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+
+/// This is a class because we need to inform the transaction about the response body being deinitialized.
+/// If the users has not called `makeAsyncIterator` on the body, before it is deinited, the http
+/// request needs to be cancelled.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+@usableFromInline final class TransactionBody: Sendable {
+    @usableFromInline let transaction: Transaction
+
+    init(_ transaction: Transaction) {
+        self.transaction = transaction
+    }
+    
+    deinit {
+        self.transaction.responseBodyDeinited()
+    }
+}
+
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+extension TransactionBody: AsyncSequence {
+    @usableFromInline typealias Element = AsyncIterator.Element
+
+    @usableFromInline final class AsyncIterator: AsyncIteratorProtocol {
+        @usableFromInline struct ID: Hashable {
+            private let objectID: ObjectIdentifier
+
+            init(_ object: AsyncIterator) {
+                self.objectID = ObjectIdentifier(object)
+            }
+        }
+
+        @usableFromInline var id: ID { ID(self) }
+        @usableFromInline let transaction: Transaction
+
+        @inlinable init(transaction: Transaction) {
+            self.transaction = transaction
+        }
+
+        deinit {
+            self.transaction.responseBodyIteratorDeinited(streamID: self.id)
+        }
+        // TODO: this should be @inlinable
+        @usableFromInline func next() async throws -> ByteBuffer? {
+            try await self.transaction.nextResponsePart(streamID: self.id)
+        }
+    }
+
+    @inlinable func makeAsyncIterator() -> AsyncIterator {
```

Comment:
We do something similar already here:
https://github.com/swift-server/async-http-client/blob/f17a47e91684541c84a0cdd8864af68e7e30c463/Sources/AsyncHTTPClient/AsyncAwait/Transaction%2BStateMachine.swift#L641-L653
This implementation is a bit more clever. It will allow the creation of multiple iterators and will only fail slightly delayed if someone tries to iterate over more than one of them. I think this is actually too clever and we should remove this eventually and just fail early during the creation of a second iterator. I'm not aware of any use case for creating unused iterators but maybe @fabianfett has some in mind.

Reply:
I'm inclined to agree with Fabian: there's no good sense in breaking the use-case that already exists, even though it'd be better defined if we did.

Reply:
Changed it back to the old behaviour. However, we now only enforce this on user provided `AsyncSequence`s.  Can also change that to use the more relaxed precondition if you want.